Fix get cache key and make package

IvanMalison opened the pull request:
Closes #110 (use py.test/tox and the test directories for unit tests) and closes #112

IvanMalison commented:
You mentioned that you are opposed to using py.test in #110 . It probably makes sense to add some type modern test runner don't you think? I have a weak preference for py.test, but nose would obviously be fine as well. Do you want me to make the bugfix in 763c78c a separate pull request from the addition of tox/a test runner? I decided to put them together because i wrote a unit test for the issue and I wanted to use a real package layout etc.

hugovk commented on setup.py:
```diff
@@ -23,10 +22,13 @@ def get_build():
     name="pylast",
     version="1.0." + get_build(),
     author="Amr Hassan <amr.hassan@gmail.com>",
+    install_requires=['pyyaml'],
```
Is this needed? Only test_pylast.py imports yaml and pyyaml is also in tests_require.

IvanMalison replied:
yep, sorry, I got a pyyaml error when I ran tox and wasn't playing close attention.

hugovk commented:
A test runner makes sense, especially for more than one test file.

Generally I'd prefer separate PRs for separate things, but the bugfix can stay inside this PR because they're kind of linked.

IvanMalison commented:
> A test runner makes sense, especially for more than one test file.

Okay, so py.test is okay?

what about the line
`@pytest.mark.parametrize('unicode_artist', [u'\xe9lafdasfdsafdsa', u'éééééééé'])`

hugovk commented:
I've not really used py.test before, that's not necessarily a bad thing, it just means I get to learn something new :)  

And that parametrize decorator looks like a really useful way to run tests with lots of different inputs and outputs, much neater than plain unittest.

So, py.test is okay for both the test runner and for writing the actual unit tests as well.

With test_pylast.py in the tests folder, the current `coverage run --source=pylast ./tests/test_pylast.py` command used by Travis CI errors: `ImportError: No module named pylast`.

Should Travis be doing something like `python setup.py install` (or `develop`) before running the tests?

Also, how to add run the unit tests with coverage?

IvanMalison commented:
I can handle both of those.

I also added tox, which can automatically create a virtualenv for the tests to run in (though we don't necessarily need to use that for travis). The command that should be used to run the tests is then tox -e py27 for python 2.6 and tox -e py34 for python 34.

jodal commented:
I can warmly recommend using tox for running the tests on Travis, as that makes it easy to reproduce a Travis test failure locally since you have the exact same setup available through `tox -e ...`.

IvanMalison commented:
@jodal Yeah, I share your preference (although weakly so), as that is what I have done when integrating with travis in the past.

One thing that I would really like is a more automatic way share a list of environments between tox and travis. It would be cool if you could use the output of tox -l to configure .travis.yml

hugovk commented on .travis.yml:
```diff
@@ -1,43 +1,35 @@
+- if [ $TOXENV == 'py27' ]; then pip install coveralls; coveralls; fi
+- if [ $TOXENV == 'py27' ]; then travis_retry pip install scrutinizer-ocular; ocular;
+  fi
+- if [ $TOXENV == 'py27' ]; then travis_retry pip install scrutinizer-ocular; ocular;
```
Duplicate line. 

Rather than just running for py27, can we have it run for all Python envs? That way we know if we have coverage where the code diverges for Py2/3.

hugovk commented on tests/test_pylast.py:
```diff
@@ -1909,54 +1908,4 @@ def test_get_recent_tracks_from_to(self):
         self.assertEqual(str(tracks[0].track.title), "Ring of Fire")
 
-
-if __name__ == '__main__':
```
I'd like to keep this main section, it's useful for running specific tests of the slow, integration tests. I know you can target from py.test, but `tox -e py27 -- project_name_here/tests/folder_name1_here/folder_name2_here/test_file_name_here.py:TestClassName.test_method_name` is a bit unwieldy.

IvanMalison replied:
py.test has a nearly identical flag (-k). It will do exactly what you want
```bash
tox -e py27 -- -k method_name should do the trick
```

hugovk replied:
OK, great! Most of main can go then, I'd to keep just this basic call so it can still be run without py.test:

```python
if __name__ == '__main__':
    unittest.main(failfast=True)
```

hugovk commented:
I ran it in my fork and it failed: https://travis-ci.org/hugovk/pylast/jobs/46284639

```
=================================== FAILURES ===================================
___________________ test_get_cache_key[troublesome_artist2] ____________________
troublesome_artist = pylast.Artist(u'B\xe9l', <Mock id='140469003402960'>)
    @pytest.mark.parametrize('troublesome_artist', [
        u'\xe9lafdasfdsafdsa', u'ééééééé',
        pylast.Artist(u'B\xe9l', mock_network())
    ])
    def test_get_cache_key(troublesome_artist):
        request = pylast._Request(mock_network(), 'some_method',
>                                 params={'artist': troublesome_artist})
tests/request_test.py:20: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pylast/__init__.py:1021: in __init__
    self.params[key] = _unicode(params[key])
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
text = pylast.Artist(u'B\xe9l', <Mock id='140469003402960'>)
    def _unicode(text):
        if sys.version_info[0] == 3:
            if type(text) in (bytes, bytearray):
                return str(text, "utf-8")
            elif type(text) == str:
                return text
            else:
                return str(text)
    
        elif sys.version_info[0] == 2:
            if type(text) in (str,):
                return unicode(text, "utf-8")
            elif type(text) == unicode:
                return text
            else:
>               return unicode(text)
E               UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 1: ordinal not in range(128)
pylast/__init__.py:3983: UnicodeDecodeError
________________________ TestPyLast.test_remove_artist _________________________
self = <tests.test_pylast.TestPyLast testMethod=test_remove_artist>
    def test_remove_artist(self):
        # Arrange
        # Get plenty of artists
        artists = self.network.get_top_artists()
        # Pick a random one to avoid problems running concurrent tests
        my_artist = choice(artists).item
        library = pylast.Library(user=self.username, network=self.network)
        library.add_artist(my_artist)
    
        # Act
>       library.remove_artist(my_artist)
tests/test_pylast.py:153: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pylast/__init__.py:2607: in remove_artist
    self._request(self.ws_prefix + ".removeArtist", False, params)
pylast/__init__.py:1320: in _request
    return _Request(self.network, method_name, params).execute(cacheable)
pylast/__init__.py:1148: in execute
    response = self._download_response()
pylast/__init__.py:1139: in _download_response
    self._check_response_for_errors(response_text)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = <pylast._Request object at 0x7fc17b6f61d0>, response = ''
    def _check_response_for_errors(self, response):
        """Checks the response for errors and raises one if any exists."""
    
        try:
            doc = minidom.parseString(_string(response))
        except Exception as e:
>           raise MalformedResponseError(self.network, e)
E           MalformedResponseError: Malformed response from Last.fm. Underlying error: no element found: line 1, column 0
pylast/__init__.py:1158: MalformedResponseError
```

IvanMalison commented:
Yes, it failed, but almost everything passed.

TestPyLast.test_remove_artist is really the only one that is surprising. Any ideas on that one. Seems like it might just be a flake. 

test_get_cache_key[troublesome_artist2] is related to #114 . I'm working on a fix right now.

hugovk commented:
`test_remove_artist` passes locally for me, so it was just a flake. That's a problem with the integration tests. They should be made less brittle, perhaps skipping or retrying for network problems. Anyway, that's for another issue.

IvanMalison commented:
hmmm so I think that I should probably fix #114 in a separate branch/pull request. I think that I will just take the test that I added out for now by reverting the test that I added and then just revert the revert later in that PR.

IvanMalison commented:
Here is my travis CI build:
https://travis-ci.org/IvanMalison/pylast/builds/46286849

IvanMalison commented:
The clonedigger command in the lint tox environment is a little broken.

hugovk commented:
Clone Digger only supports Python 2, but it didn't fail the builds:

https://travis-ci.org/pylast/pylast/builds/46051470

IvanMalison commented:
The build you linked isn't for my PR. Also it is only the lint build that will fail: https://travis-ci.org/IvanMalison/pylast/jobs/46286855

All the other builds don't run the clonediggers stuff.

hugovk commented:
Yep, https://travis-ci.org/pylast/pylast/builds/46051470 is for an older build. Clone Digger was called in the `after_script` for each Python version. Errors for Python 3 and PyPy didn't cause any build failures.

But I see your point, the lint step in https://travis-ci.org/IvanMalison/pylast/jobs/46286855 isn't run in the other builds. 

By the way, for my CI builds I set concurrent builds to 1 because the integration tests are modifying the state of Last.fm. That can cause a test from a parallel build to fail because something external has changed during its run. (Another reason for proper unit tests!)

hugovk commented:
I've merged this into a new develop branch on pylast and all pass except the lint one, which is taken care of in another PR.

https://travis-ci.org/pylast/pylast/builds/46300182

I'll merge it to master. Thanks for all this!
